In [214]:
#import the necessary librRIES
import tensorflow as tf
from tensorflow import keras
from tensorflow.keras.preprocessing.image import ImageDataGenerator



In [215]:
dataset_path="PetImages"

In [216]:
datagen=ImageDataGenerator(
    #Normalize pixels
    rescale=1./255,
    #data augmentation
    shear_range=0.2,
    zoom_range=0.2,
    horizontal_flip=True,
    #split dataset
    validation_split=0.2
)

In [217]:
# LOad training Data
#Training data
train_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='training'
)

Found 19968 images belonging to 2 classes.


In [218]:
# LOad training Data
#Training data
val_data=datagen.flow_from_directory(
    dataset_path,
    target_size=(128,128),
    batch_size=32,
    class_mode='binary',
    subset='validation'
)

Found 4991 images belonging to 2 classes.


In [219]:
print(train_data.class_indices)

{'Cat': 0, 'Dog': 1}


In [220]:
model=keras.Sequential()
# 1.Simple Images ---> Few Layers
 
# ex: handwritten numbers
 
# (28*28)
 
# 2 conc 2 pool
 
# 2.Medium Complexity
 
# Cats vs Dogs
 
# (128*128)
 
# (224*224)
 
# 3 Conv + 3 Pool.
 
 
 

In [221]:
#first convolutional layer
model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu',
        input_shape=(128,128,3)
    )
)

In [222]:
#first pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)


    )
)

In [223]:
#second convolutional layer
model.add(
    keras.layers.Conv2D(
        filters=32,
        kernel_size=(3,3),
        activation='relu'
    )
)

In [224]:
#second pooling layer
model.add(
    keras.layers.MaxPooling2D(
        pool_size=(2,2)
    )
)

In [225]:
model.add(keras.layers.Flatten())

In [226]:
model.add(
    keras.layers.Dense(units=128,activation='relu')
)

In [227]:
model.add(
    keras.layers.Dense(
        units=1,
        activation='sigmoid'
    )
)

In [228]:
#View model architecture
model.summary()

Model: "sequential_11"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ conv2d_30 (Conv2D)              │ (None, 126, 126, 32)   │           896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_28 (MaxPooling2D) │ (None, 63, 63, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ conv2d_31 (Conv2D)              │ (None, 61, 61, 32)     │         9,248 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ max_pooling2d_29 (MaxPooling2D) │ (None, 30, 30, 32)     │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ flatten_10 (Flatten)            │ (None, 28800)          │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 128)            │     3,686,528 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_21 (Dense)                │ (None, 1)              │           129 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 3,696,801 (14.10 MB)

 Trainable params: 3,696,801 (14.10 MB)

 Non-trainable params: 0 (0.00 B)

(None,126,126,Channels)
(Batch size,Height,Width,NO.of Filters)
### 1 Conv2D
Input = 128*128*3

filter = 3*3

no.of filters = 32

output sizze formula:

output=((input-kernel)/(strides))+1

Strides=1

output=((128-3)/1)+1

      =126

height : 126

width : 126

Channels------>32

final output -----> (None,126,126,32)

kernel

formula=(kh * kw * InputChannels + 1)*Filters

kernel = 3*3

InputChannels : 3

Filters : 32

params = (3*3*3 + 1)*32

       = (27 + 1)*32

       = (28 * 32)

       = 896

### Pooling2D

MaxPooling2D((2,2))

Pooling = No learning , No weights

(only reduces size)

Input = 126*126*32

pool=2*2

Formula : 126/2

       =63

output : (63*63*32)

Final : (None,63,63,32)

### Second Conv2D_1

Input from pooling : (None,63*63*32)

kernel : 3*3

output : 63-3+1
        
      =61

(None,61,61,32)

Param = 9248

kernels = (3*3*32 + 1)*32

        =(288 + 1)*32

        =289 * 32

        =9248
## second Pooling

(None,30,30,32)

Params = 0

Input=(61,61,32)

Params = 0

Pooling : 2*2

Floor division : 61/2 = 30

Final : (None,30,30,32)

No Parameters.



### Flatten

Input : 30*30*32

output : 30*30*32 = 28800

Flatten : It will convert 3D into 1D vector .. (Flatten only reshapees..)

## Dense Layer

Input = 28800

Neurons  = 128

Parameters = (Input * Neurons) + Bias

    = ( 28800 * 128 ) + 128

    =3686400 + 128

    =3686528

## Output Layer

Input = 128

Output = 1

Output shape (None,1)

-----> Cat or Dog

Paramter Calculation:

        =128 * 1 + 1
        
        =129

Input -----> Conv1 --> Pool1 ----> Conv2 ---->Pool2------> Flatten -----> Dense ----> Output


In [229]:
import tensorflow as tf
print(tf.config.list_logical_devices())

[LogicalDevice(name='/device:CPU:0', device_type='CPU')]


In [230]:
#Compile CNN
model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

In [231]:
model.fit(train_data,epochs=5,validation_data=val_data)

Epoch 1/5
221/624 ━━━━━━━━━━━━━━━━━━━━ 2:41 401ms/step - accuracy: 0.5686 - loss: 0.7239

c:\Users\tarun\AppData\Local\Programs\Python\Python313\Lib\site-packages\PIL\TiffImagePlugin.py:950: UserWarning: Truncated File Read
  warnings.warn(str(msg))


624/624 ━━━━━━━━━━━━━━━━━━━━ 335s 534ms/step - accuracy: 0.6357 - loss: 0.6462 - val_accuracy: 0.6934 - val_loss: 0.5861
Epoch 2/5
624/624 ━━━━━━━━━━━━━━━━━━━━ 328s 525ms/step - accuracy: 0.7127 - loss: 0.5641 - val_accuracy: 0.7467 - val_loss: 0.5224
Epoch 3/5
624/624 ━━━━━━━━━━━━━━━━━━━━ 288s 461ms/step - accuracy: 0.7557 - loss: 0.5000 - val_accuracy: 0.7347 - val_loss: 0.5290
Epoch 4/5
624/624 ━━━━━━━━━━━━━━━━━━━━ 154s 247ms/step - accuracy: 0.7789 - loss: 0.4642 - val_accuracy: 0.7842 - val_loss: 0.4607
Epoch 5/5
624/624 ━━━━━━━━━━━━━━━━━━━━ 164s 263ms/step - accuracy: 0.8020 - loss: 0.4262 - val_accuracy: 0.7768 - val_loss: 0.4732


In [232]:
loss,accuracy=model.evaluate(val_data)
print("accuracy:",accuracy)

156/156 ━━━━━━━━━━━━━━━━━━━━ 20s 130ms/step - accuracy: 0.7734 - loss: 0.4687
accuracy: 0.7733920812606812
